lib/ukdebug: Add `README.md` to `ukdebug` (#1492)
unikraft/unikraft: lib/ukdebug/README.md

# `ukdebug`: Unikraft's debugging features

This library contains features to debug Unikraft.

## GDB stub

The GDB stub allows remotely debugging Unikraft with GDB using the [GDB Remote Serial Protocol (RSP)](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Remote-Protocol.html).
The debugging process involves two machines.

> A development machine (the GDB host) that runs GDB and implements the client side of the RSP.

Reviewer: ```suggestion
A development machine (the GDB host) runs GDB and implements the client side of the RSP.
```

Author: I disagree with this. IMO the `that` makes sense in this sentence.

Reviewer: It's not a full sentence tho. I don't have a hard preference here however.

Author: OK you have convinced me :)  I removed the `that`.

The other machine runs Unikraft (the target), which implements the GDB stub and serves requests issued by the host.
The two sides communicate over a serial connection.
With this setup, the machine running Unikraft can be a local VM or a physically remote machine.

```plain
   Development machine                     Unikraft
  ┌─────────────┐                         ┌─────────────┐
  │             │    Serial connection    │             │
  │  GDB Host ◄─┼─────────────────────────┼─► GDB Stub  │
  │             │                         │             │

>   └─────────────┘                         └─────────────┘

Reviewer 2: Use ASCII characters for this pls

Reviewer 3: @StefanJum can you please elaborate on the rationale behind this requirement?

Reviewer 2: (As discussed in the maintainers meeting)
1. We can be sure ASCII characters are rendered properly everywhere, no matter the layout/language etc.
2. The README files will eventually be part of the documentation website, were we already use the ASCII-only convention, so it makes sense for consistency reasons.

Author: Fair points. I switched the graphic to ASCII :+1:

```

The stub runs after every debug trap, meaning the stub executes in an exception context.

> A loop inside the stub waits to receive packets from the host and act upon them.

Reviewer: ```suggestion
A loop inside the stub waits to receive packets from the host and acts upon them.
```

Author: This sentence was also phrased ambiguously. I meant to say that inside of the loop packets are received and acted upon. I switched from `act` to `to act` to hopefully make this more clear.

> Once the stub is finished handling a packet, it will send the response to the host and wait for the next packet.

Reviewer: ```suggestion
Once the stub finishes handling a packet, it will send the response to the host and wait for the next packet.
```

Author: I prefer `is finished` here. Saying `Once the stub finishes ...` sounds like the stub sends the response at the same time as it finishes handling the response while, in fact, it sends the response after.

Reviewer: The original meaning stays intact due to the usage of the _once_ conjunction, which here means _as soon as_. I don't have a hard preference for either one of the wordings, however.

Author: OK I'm keeping it as it is now.

Some packets instruct the stub to leave the loop and resume execution of the kernel.

The GDB stub uses a serial console device to communicate with the host.
This can be either the same console that is used for standard output, or a separate console dedicated for debugging.
If the GDB stub and standard output use the same console, Unikraft's standard output is printed inline with GDB's output in the GDB CLI.

### Kernel parameters

Two kernel parameters control the GDB stub.

- `debug.gdbcon=<id>`: ID of the console device that is used to communicate with the GDB host.
  If there's only a single console on the system, its ID is `0`, so `debug.gdbcon=0` should be used to select it.
  The GDB stub is only enabled at runtime if `debug.gdbcon` is set.
- `debug.gdbmode=[0, 1, -1]`: A tristate value controlling the mode of the GDB stub.

>   The three states for the tristate are "shared" (`1`), "not shared" (`0`) and "auto" (`-1`).

Reviewer: ```suggestion
  The three states are "shared" (`1`), "not shared" (`0`) and "auto" (`-1`).
```

E.g., as an alternative to the above, you could configure COM1 and COM2 in kconfig and then connect the GDB stub to COM2 and keep kernel output on COM1 (`debug.gdbcon` is `1` this time):

```bash
qemu-system-x86_64 -kernel build/app-helloworld_qemu-x86_64 \
                   -cpu max \
                   -display none \
                   -serial stdio \
                   -serial tcp::1235,server,nowait \
                   -append "build/app-helloworld_qemu-x86_64 debug.gdbcon=1 --"
```

But let's stick to the initial example and assume that only COM1 is configured and that `debug.gdbcon=0`.
Start the VM and in another terminal run GDB like:

```bash
gdb build/app-helloworld_qemu-x86_64.dbg
```

There are two things to note here.
First, make sure to use the `.dbg` image, as it contains symbol and debug information.
GDB won't find any symbols without it.
Second, make sure you're using a  GDB binary compiled for the target architecture.
If you're cross-compiling, you must use the GDB binary provided by the cross-compiling toolchain.

Run the command `target remote localhost:1235` in the GDB CLI to connect to the GDB stub on TCP socket `localhost:1235`.

> GDB should manage to connect to the VM and you should see where in the kernel execution was stopped.

Reviewer: ```suggestion
GDB should manage to connect to the VM and you should see where in the kernel execution it was stopped.
```

Author: I meant to refer to the `execution` as a single noun here, not `kernel execution` as your change suggests. This makes me think I need to reword this sentence to make it more clear. See if you like the update.

# `ukdebug`: Unikraft's debugging features

This library contains features to debug Unikraft.

## GDB stub

> The GDB stub allows remotely debugging Unikraft with GDB using a feature of

Reviewer 3: > with GDB using a feature of GDB called the remote protocol

A way more suitable for technical writing would be  "with GDB, using the [GDB Remote Serial Protocol](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Remote-Protocol.html)".

> GDB called the remote protocol. The way this works is you run an instance of

Reviewer 3: > The way this works is you run an instance of GDB on your local computer that you connect to Unikraft.

Ditto, this can be improved to sound a bit less informal:

"The debugging process involves two machines. A development machine (the GDB host) that runs GDB, which implements the client side of the RSP. The other machine runs Unikraft (the target) implements the GDB stub, which serves the requests issued by the host. The two sides communicate over a serial connection. With this setup, the machine running Unikraft can be a local VM or a physically remote machine."

GDB on your local computer that you connect to Unikraft. This instance of
GDB is called the GDB host, and it implements one side of the GDB remote
protocol. The other side is implemented by the GDB stub in Unikraft and the
two sides talk to each other over a serial connection. With this setup, the
machine running Unikraft can be a local VM or a physically remote machine.

```plain

>   Your machine (any OS)                Other machine (Unikraft)

Reviewer 3: "Your machine (any OS)" -> "Development machine"

"Other machine (Unikraft)" -> "Unikraft".

  ┌─────────────┐                         ┌─────────────┐
  │             │    Serial connection    │             │
  │  GDB Host ◄─┼─────────────────────────┼─► GDB Stub  │
  │             │                         │             │
  └─────────────┘                         └─────────────┘
```

> Unikraft's GDB stub lives in `gdbstub.c` with architecture-specific code in

Reviewer 3: This sentence can be dropped. In principle it's best to avoid documenting implementation as much as possible as such documentation needs to be maintained along with the code. In practice this means it will eventually rot as people tend not to update documentation.

`arch/$(UK_ARCH)/gdbsup.c`.

The stub runs after every debug trap (`UKARCH_TRAP_DEBUG`) meaning the stub

> executes in an exception context. There is a loop inside the stub that waits to

Reviewer 3: >  There is a loop inside the stub that waits

"A loop inside the stub waits..."

receive packets from the host and act upon them. Once the stub is finished
handling a packet, it will send the response to the host and wait for the next
packet. Some packets instruct the stub to leave the loop and resume execution
of the kernel.

> The GDB stub uses a serial console device (`struct uk_console`) to communicate

Reviewer 3: Please remove (`struct uk_console`).

with the host. This device is referred to as the "backing console" because it
backs the connection to the host. When the stub is initially set up, it
retrieves a reference to the console device for the backing console from

> `ukconsole` and clears the `UK_CONSOLE_FLAG_STDIN` and `UK_CONSOLE_FLAG_STDOUT`

Reviewer 3: The whole description of flags (`UK_CONSOLE_FLAG_STDIN` and `UK_CONSOLE_FLAG_STDOUT`) should not be here for maintainability reasons described above. Same for the implementation of backing / virtual console and how the stub takes over.

Additionally to the above these details only adds confusion to a user that is not aware of the implementation (consider the case that the reader's intention is to debug their application).

As the next paragraph is also based on this information, I would replace the two with a more high-level description:

"The GDB stub uses a serial console device to communicate with the host. This can be either the same console that is used for standard output, or a separate console dedicated for debugging. If a single console is used, Unikraft's standard output is printed inline with GDB's output in the GDB CLI.
"

flags for that device.

Clearing the flags gives the GDB stub exclusive access to the backing console.
If the GDB stub didn't have exclusive access to the backing console, any kernel
output would interfere with the GDB communications and potentially break the
connection. However, even on systems that only feature a single console, we
would still like to see kernel output. For this reason, the GDB stub has a mode
to share the backing console between the GDB communications and kernel output.
This is done by re-routing all kernel output and sending it to the GDB host
wrapped in a special packet. The GDB host then prints the output that it
receives to `stdout`, inline with the normal output from the GDB CLI.

### Kernel parameters

Two kernel parameters control the backing console and the mode of the GDB stub.

> - `debug.gdbcon`: This is the console ID of the console device from `ukconsole`

Reviewer 3: `debug.gdbcon=<id>`

> - `debug.gdbcon`: This is the console ID of the console device from `ukconsole`

Reviewer 3: >  This is the console ID of the console device from `ukconsole`  that should be used as the backing console.

"The ID assigned to the console GDB should connect to."

>   that should be used as the backing console; it's a 16-bit unsigned integer.

Reviewer 3: >  it's a 16-bit unsigned integer.

For reasons described earlier, I would avoid documenting the implementation in such detail.

  If there's only a single console on the system, its ID is `0`, so
  `debug.gdbcon=0` should be used to select it. The GDB stub is only enabled at
  runtime if `debug.gdbcon` is set.

> - `debug.gdbmode`: This is a tristate value controlling the mode of the GDB

Reviewer 3: "A tristate value controlling..."

> - `debug.gdbmode`: This is a tristate value controlling the mode of the GDB

Reviewer 3: > `debug.gdbmode:`

`debug.gdbmode=[0, 1, -1]:`.

> The three states for the tristate are "shared", "not shared", and "auto". Any negative value is interpreted to mean  "auto", a value of zero is interpreted to mean "not shared" and a positive value is interpreted as "shared".

In retrospective we should had used keywords for these instead of numbers. 

Either way, by defining the syntax as above you can now be more terse: "The three states for the tristate are "shared" (`0`), "not shared" (`1`), and "auto" (`-1`).". The default is "auto".

Although technically correct, I would avoid documenting the any positive / negative value behavior as it makes no sense to use values like `3` or `-6` anyway.

>   stub. Its type is a signed integer. The three states for the tristate are

Reviewer 3: > Its type is a signed integer

Again this is unnecessary.

  "shared", "not shared", and "auto". Any negative value is interpreted to mean
  "auto", a value of zero is interpreted to mean "not shared" and a positive

>   value is interpreted as "shared". "Shared" mode is useful when the system

Reviewer 3: > "Shared" mode is useful when the system only has a single console, its operation was described above. 

"The "shared" mode...".

> If you don't include this kernel parameter, "auto" will be used.

You can skip that by adopting the sentece in comment above that states "The default is "auto"."

  only has a single console, its operation was described above. If you don't

>   include this kernel parameter, "auto" will be used. With "auto" the selection

Reviewer 3: > With "auto" the selection between "shared" and "not shared" mode is made based on a simple policy. If the device used as the backing console had any flags set before being taking over, then "auto" will select "shared" mode. Otherwise, if the backing console didn't have flags set, "auto" will select "not shared" mode.

This should be rephrased to more high-level terms (no implementation-specific description like flags).

  between "shared" and "not shared" mode is made based on a simple policy. If
  the device used as the backing console had any flags set before being taking
  over, then "auto" will select "shared" mode. Otherwise, if the backing
  console didn't have flags set, "auto" will select "not shared" mode.

### QEMU & GDB example

What follows is a concrete example of using the GDB stub with QEMU and GDB. For
starters, configure your project to enable the GDB stub and a single console

> device (COM1 on x86, PL011 on Arm). Note that the GDB stub depends on

Reviewer 3: > (COM1 on x86, PL011 on Arm)

Arm can use other drivers, so maybe that's unnecessary.

```bash
qemu-system-x86_64 -kernel build/app-helloworld_qemu-x86_64 \
                   -cpu max \
                   -display none \
                   -serial tcp::1235,server,nowait \
                   -append "build/app-helloworld_qemu-x86_64 debug.gdbcon=0 --"
```

The only two interesting options here are `-serial` and `-append`.
`-serial tcp::1235,server,nowait` makes QEMU create a TCP socket on
`localhost:1235` that's connected to COM1 inside the VM. This TCP socket will
be used to connect the GDB host to the serial console in the VM. `-append`
specifies the kernel command line that's passed to the kernel at runtime.
Passing the name of the image as the first parameter is an Unikraft convention.
If the first parameter is a kernel parameter, this kernel parameter will be
dropped, so be careful. With `debug.gdbcon=0`, we tell the GBD stub to use
`con0` as the backing console. On x86 this is COM1 and it's connected to
`localhost:1235`.

You can use `-serial stdio` instead of `-serial tcp::1235,server,nowait` to
see if the GDB stub is enabled and set up correctly. If you can run above
command with `-serial stdio`, you should see some output that ends with

> something like the below (assuming you have debug messages enabled):

Reviewer 3: > assuming you have debug messages enabled)

assuming verbosity set to >= `info`

[    0.104944] Info: [libukconsole] <console.c @  176> Registered con1: GDB virtual console, flags: --
$S05#b8
```

The last line is the first packet that the GDB stub sends to the GDB host.
Switch back to `-serial tcp::1235,server,nowait` and you'll stop seeing any
output. Using `-serial stdio` can be useful when you're unsure if your
configured kernel even boots.

Note that the `-serial` flag is position-sensitive on x86. COM1 is connected
to whatever the first occurrence (from left to right) of the `-serial` flag
specifies. COM2 is configured based on the second occurrence of `-serial`, and
so forth. E.g., as an alternative to the above, you could configure COM1 and
COM2 in kconfig and then connect the GDB stub to COM2 and keep kernel output on
COM1 (`debug.gdbcon` is `1` this time):

```bash
qemu-system-x86_64 -kernel build/app-helloworld_qemu-x86_64 \
                   -cpu max \
                   -display none \
                   -serial stdio \
                   -serial tcp::1235,server,nowait \
                   -append "build/app-helloworld_qemu-x86_64 debug.gdbcon=1 --"
```

But let's stick to the initial example and assume that only COM1 is configured

> and that `debug.gdbcon=0`. Start the VM and in another terminal run GDB like

Reviewer 3: > run GDB like this:

"run GDB like:"

this:

```bash
gdb build/app-helloworld_qemu-x86_64.dbg
```

> There are two things to note here. First, don't forget the `.dbg` extension.

Reviewer 3: > First, don't forget the `.dbg` extension. GDB won't find any symbols without it.

"Make sure you use the `.dgb` image, as it contains object symbols."

> Second, make sure you're using the right GDB binary. E.g., if you're developing on an x86 system and the kernel is compiled for Arm, you need to use GDB for Arm, not the stock GDB of your system.

I would avoid making specific assumptions about the host here, as with Mac it's also common to be on Arm and cross-compile for `x86_64`, so: "Second, make sure you're using a GDB binary compiled for the target architecture. If you're cross-compiling you must use the GDB binary provided by the cross-compiling toolchain".

GDB won't find any symbols without it. Second, make sure you're using the right
GDB binary. E.g., if you're developing on an x86 system and the kernel is
compiled for Arm, you need to use GDB for Arm, not the stock GDB of your system.

> Run the command `target remote localhost:1235` in the GDB REPL to connect to

Reviewer 3: I dont't think most people know what a REPL is, please use "CLI" or "interface".

the GDB stub on TCP socket `localhost:1235`. GDB should manage to connect to
the VM and you should see where in the kernel execution was stopped. E.g., it
might look like this:

```plain
(gdb) tar remote :1235
Remote debugging using :1235
0x0000000000120e02 in gdb_entry (ctx=<optimized out>) at /home/thasso/Work/oss-unikraft/lib/ukdebug/gdbstub.c:1247
1247    }
```

You are now connected to the GDB stub that runs inside Unikraft. All the
usual GDB commands like `break`, `backtrace`, or `continue` should work. Since
we only configured a single console and didn't set `debug.gdbmode` on the
command line, the GDB should automatically share the backing console for kernel
output. If you `break main` and `continue`, you should see all kernel output
printed inline with the other output from GDB.

### Debugging the GDB stub

> This is guidance for anyone working on the GDB stub who needs to debug it.

Reviewer 3: > This is guidance for anyone working on the GDB stub who needs to debug it.

"This section provides information for developers who need to debug Unikraft's GDB stub."
